File: HW6/gtuset.cpp
//Implementations for GTUSET class
#include "gtuset.h"

#include <memory>

namespace TurkerTercanHW6
{
	template<class T>
	//Default Constructor
	GTUSet<T>::GTUSet()
	{
		this->capacity = 1;
		std::shared_ptr<T> temp(new T[this->capacity]);
		this->data = temp;
		this->used = 0;
	}
	
	//Returns true if it is empty
	template<class T>
	bool GTUSet<T>::empty() const noexcept
	{
		if(this->used != 0)
			return false;
		else
			return true;
	}
	
	//Returns used
	template<class T>
	int GTUSet<T>::size() const noexcept
	{
		return this->used;
	}
	
	//Returns capacity
	template<class T>
	int GTUSet<T>::max_size() const noexcept
	{
		return this->capacity;
	}
	
	//Insert an element to container if it not already in the container.
	template<class T>
	void GTUSet<T>::insert(const T& object) noexcept(false)
	{
		bool inSet = false;
		for(auto i = this->begin(); i != this->end(); ++i)
		{
			if(*i == object)
			{
				throw std::invalid_argument("Error: You can not insert an element which is already in the set!");
				inSet = true;
			}
		}
		//If the capacity is not reached, attach the element to the end()
		if(this->size() != this->max_size() && !inSet)
		{
			*(this->end()) = object;
			this->used++;
		}
		
		//If the capacity is reached, double the capacity and attach the element to the end()
		else if(this->size() >= this->max_size() && !inSet)
		{
			this->capacity = this->capacity * 2;
			std::shared_ptr<T> temp(new T[this->capacity]);
			int i = 0;
			for( auto iter = this->begin(); iter != this->end(); ++iter, ++i) //Transferring all elements in shared_ptr to new one
			{
				temp.get()[i] = *(iter);
			}
			this->data = temp;
			*(this->end()) = object;
			this->used++;
		}
	}
	//Erase the first element if there is one 
	template<class T>
	void GTUSet<T>::erase(const T& object) noexcept
	{
		bool inSet = false;
		for( auto i = this->begin(); i != this->end(); i++ )
		{
			if(*i == object)
			{
				for( auto j = i; j != this->end(); j++ )
				{
					*j = *(j.get() + 1);
				}
				inSet = true;
			}
		}
		if(inSet == 1)
			this->used--;
	}
	
	//Delete all elements in container
	template<class T>
	void GTUSet<T>::clear() noexcept
	{
		this->used = 0;
		this->capacity = 1;
		std::shared_ptr<T> temp(new T[this->capacity]);
		this->data = temp;
	}
	
}

File: HW8/src/hw08/GTUSet.java
package hw08;

import java.lang.reflect.Array;

/**
 * This is a simple set class which is extends from GTUContainer abstract class
 * 
 * @author Türker Tercan
 */
public class GTUSet<T> extends GTUContainer<T>{

	/**
	 * Set constructor to sets the its capacity to 1 
	 * and creates a empty array of capacity's size
	 * 
	 * @param _type is the type of array that is needed to be instantiated
	 */
	public GTUSet(Class<T[]> _type)
	{
		used = 0;
		capacity = 1;
		genericType = _type;
		T[] temp =  genericType.cast(Array.newInstance(genericType.getComponentType(), capacity));
		this.data = temp;
	}
	
	@Override
	public boolean empty()
	{
		if( used == 0 )
			return true;
		else
			return false;
	}
	
	@Override
	public int size()
	{
		return used;
	}
	
	@Override
	public int max_size()
	{
		return capacity;
	}
	
	@Override
	public void insert( T element ) throws IllegalArgumentException
	{
		if( this.contains(element) )
		{
			throw new IllegalArgumentException("Error : You can not insert an element which is already in the set!\n");
		}
		
		else
		{
			if( used < capacity )
			{
				data[used++] = element;
			}
			else
			{
				capacity = capacity * 2;
				GTUIterator<T> p = this.iterator();
				T[] temp = genericType.cast(Array.newInstance(genericType.getComponentType(), capacity));
				for( int i = 0; i < used; i++ )
				{
					temp[i] = p.next();
				}
				this.data = temp;
				this.data[used++] = element;
			}
		}
	}
	
	@Override
	public void erase( T element )
	{
		if(this.contains( element ))
		{
			GTUIterator<T> p = this.iterator();
			while( p.hasNext() )
			{
				if( p.next() == element)
				{
					int i = p.getPosition() - 1;
					while( p.hasNext() )
					{
						data[i++] = p.next();
					}
				}
			}
			used--;
		}
		else
			System.err.printf("%s could not be found in the set\n", element);
	}
	
	@Override
	public void clear()
	{
		used = 0;
		capacity = 1;
		T[] temp = genericType.cast(Array.newInstance(genericType.getComponentType(), capacity));
		this.data = temp;
	}
	
	@Override
	public GTUIterator<T> iterator()
	{
		return new GTUIterator<T>(data);
	}
	
	@Override
	public boolean contains( Object o )
	{
		GTUIterator<T> temp = this.iterator();
		while( temp.hasNext() )
		{
			if( temp.next() == o)
				return true;
		}
		return false;
	}
	
	@Override
	public String toString()
	{
		GTUIterator<T> p = this.iterator();
		String[] totalSet = new String[used+1];
		totalSet[0] = String.format("Printing %s size = %d, max_size = %d\n", this.getClass(), used, capacity);
		
		for(int i = 1; i < totalSet.length; i++)
		{
			totalSet[i] = String.format("%d. element: %s\n", i, p.next());
		}
		
		StringBuilder build = new StringBuilder();
		for(int i = 0; i < totalSet.length; i++)
		{	
			build.append(totalSet[i]);
		}
		return build.toString();
	}
}
